### PR TITLE
Tests for default score, probability floor, and top score tie-breaking

### PR DESCRIPTION
**Pull request by aarsilv**
_Eppo Internal:_
🎟️ **Ticket:** [FF-2996 - Update shared bandit tests to handle more edge cases](https://linear.app/eppo/issue/FF-2996/update-shared-bandit-tests-to-handle-more-edge-cases)

This PR adds a new bandit that powers a new test file with two test cases that check three edge cases:
- Default score is correctly applied
- Probability floor is correctly applied
- Top-scoring actions are tie-broken by name for weighting

### CODE REVIEW

File: ufc/bandit-flags-v1.json [hunk 1]
```diff
@@ -201,6 +201,31 @@
                 }
             ],
             "totalShards": 10000
+        },
+        "car_bandit_flag": {
```
Comment (aarsilv) [inline ufc/bandit-flags-v1.json:205]: A new bandit so we can safely play with the parameters without messing up existing test cases. For simplicity, we assign 100% of the traffic to this bandit

File: ufc/bandit-models-v1.json [hunk 2]
```diff
@@ -106,6 +106,31 @@
...
       }
     },
+    "car_bandit": {
+      "banditKey": "car_bandit",
+      "modelName": "falcon",
+      "updatedAt": "2023-09-13T04:52:06.462Z",
+      "modelVersion": "v456",
+      "modelData": {
+        "gamma": 1.0,
+        "defaultActionScore": 5.0,
```
Comment (aarsilv) [inline ufc/bandit-models-v1.json:116]: Non-zero default score

File: ufc/bandit-models-v1.json [hunk 3]
```diff
@@ -106,6 +106,31 @@
...
+        "actionProbabilityFloor": 0.2,
```
Comment (aarsilv) [inline ufc/bandit-models-v1.json:117]: Non-zero probability floor

File: ufc/bandit-models-v1.json [hunk 4]
```diff
@@ -106,6 +106,31 @@
...
+        "coefficients": {
+          "toyota": {
+            "actionKey": "toyota",
+            "intercept": 1.0,
+            "actionNumericCoefficients": [{
+              "attributeKey": "speed",
+              "coefficient": 1,
```
Comment (aarsilv) [inline ufc/bandit-models-v1.json:124]: Makes it easy to score actions how we want

File: ufc/bandit-tests/test-case-car-bandit.json [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                "numericAttributes": {},
+                "categoricalAttributes": {}
+            },
+            {
+              "actionKey": "toyota",
+              "numericAttributes": {"speed":  1000},
+              "categoricalAttributes": {}
+            }
+        ],
+        "assignment": {"variation": "car_bandit", "action": "dodge"}
```
Comment (aarsilv) [inline ufc/bandit-tests/test-case-car-bandit.json:25]: The above actions shuffle to: Mercedes, Dodge, Toyota

Without minimum probability floors, Toyota gets most of the weight, and that's what gets assigned. However, with a probability floor, the other actions each get a non-trivial shard range, so it is not assigned.

If the probability floor is not normalized (divided by the number of actions), then the first action, Mercedes, will be assigned too many shards and get assigned.

Only when it's non-zero and normalized does Anastasia get correctly assigned the middle action: Dodge.

File: ufc/bandit-tests/test-case-car-bandit.json [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+              "numericAttributes": {},
+              "categoricalAttributes": {}
+            },
+            {
+              "actionKey": "ford",
+              "numericAttributes": {},
+              "categoricalAttributes": {}
+            }
+        ],
+        "assignment": {"variation": "car_bandit", "action": "ford"}
```
Comment (aarsilv) [inline ufc/bandit-tests/test-case-car-bandit.json:47]: The above actions shuffle to: Toyota, Honda, Ford

If the default action score is ignored and the unknown (non-Toyota) actions get scored `0`, Toyota takes up too many shards and that is what is assigned.

With the default score at play, the two unknown actions are top-scoring and get most of the shards. However, because of the top scoring action weight of 1 - all other action weights, the top scoring action gets the most.

When not tie-broken, implementations like JavaScript that iterate over them in order would declare Honda the top-scoring action, and it would end up getting assigned. However, when breaking ties by action name, Ford takes the prize and gets assigned.

Thus, for this test to pass with Ford being assigned to Bob, the default action scores need to be respected, and top-scoring actions need to be tie-broken by name.

Comment (typotter) [reply ufc/bandit-tests/test-case-car-bandit.json:47]: hot dang, this is great test crafting; thanks for the detailed comment.

1) I thought the decision in the end was to not worry about the precision of tiebreaking the top score by name, but it really is a simple change required in the other SDKs to support this, and it eliminates the ambiguity (while arbitrary, it will at least be consistent)

Would adding a `remarks` field to the `assignment` object in the JSON break parsing in the SDK test suites? It would be really helpful to have at least some of your detailed explanations of the test case captured in the file.

File: ufc/bandit-flags-v1.json [hunk 7]
```diff
@@ -218,6 +243,14 @@
                 "variationValue": "banner_bandit"
             }
         ],
+        "car_bandit": [
```
Comment (typotter) [inline ufc/bandit-flags-v1.json:246]: missing `car_bandit` in `banditReferences`. Hopefully we can fully deprecate and remove `bandits` at some point in the future... maybe when everyone is on UFCv2 (whatever that may look like)

**Review by typotter — APPROVED** [review]
Approving, but you'll need to update `banditReferences` before merging to keep the SDKs on the "new hotness" from failing the tests.

File: ufc/bandit-tests/test-case-car-bandit.json (hunk 6 above)
Comment (aarsilv) [reply ufc/bandit-tests/test-case-car-bandit.json:47]: I love the `remarks` idea, and I don't think it will break anything. I was just lamenting how JSON doesn't have comment support.

Good spot on `banditReferences` 🤦 

You're right about the initial decision to not worry about tie-breaking; I had forgotten that. I think we do update the SDKs it was really easy to do for JS and may matter in situations of default score (e.g,. a cold-started bandit where everything starts with the default score)

File: ufc/bandit-flags-v1.json (hunk 7 above)
Comment (aarsilv) [reply ufc/bandit-flags-v1.json:246]: GREAT catch! I'm embarassed I forgot that 🤦 

Added to `car_bandit` to `banditReferences` and also fixed up the incorrect allocation keys in there as well.

File: ufc/bandit-tests/test-case-car-bandit.json (hunk 6 above)
Comment (aarsilv) [reply ufc/bandit-tests/test-case-car-bandit.json:49]: Ok I added as a `comment` field to each test case subject. Unfortunately no easy way to do multi-line in JSON so they are looooong lines of text. But I used the pyramid principle!
